framework/scripts/find_hung_process.py

```diff
@@ -231,129 +283,244 @@ def generate_traces(job_num, application_name, num_hosts, num_threads):
+                tmp_trace = self.host_and_frames_re.search(trace)
+                if tmp_trace:
+                    host_pid = tmp_trace.group(1)
+                    trace = tmp_trace.group(2)
+
+                # If the user requested to save only the last few lines, do that here
+                if num_lines_to_keep:
+                    tmp_trace = last_lines_regex.search(trace)
+                    if tmp_trace:
+                        trace = tmp_trace.group(0)
+
+                unique = ''
+                for back_trace in self.unique_stack_traces:
+                    if self.compare_traces(trace, back_trace):
+                        unique = back_trace
+
+                if unique == '':
+                    self.unique_stack_traces[trace] = [host_pid]
+                else:
+                    self.unique_stack_traces[unique].append(host_pid)
+
+            # Figure out if we need to search for divergence up or down from here
+            current_result_diverged = len(self.unique_stack_traces) > 1
+            if current_result_diverged:
+                bin_search_index.set_direction(-1)
+                print("- diverged...")
+            else:
+                bin_search_index.set_direction(1)
+                print("- converged...")
+
+            """
+            See if we've for sure narrowed in on the stack where the divergence first occurs.
+            We'll only know for sure if we are testing adjacent indicies. Once we are at
+            that step though, there are four possibilities:
+
+            Case 1: Current round is a divergence - return the current result
+
+            Case 2: Last iteration was a divergence, current round is a convergence - return the last result
+            Case 3: Both the current and last iteration are converged - No divergence at! (return the last result)
+
+            case 4: All passes are converged!
+            """
+            if (abs(num_lines_to_keep - last_index) <= 1):
+                # Case 1 & 4:
+                if current_result_diverged or last_diverged_result == {}:
+                    self.reportResults(self.unique_stack_traces, num_lines_to_keep)
+                # Cases 2 & 3:
+                else:
+                    self.reportResults(last_diverged_result, last_diverged_index)
+                return
+
+            last_index = num_lines_to_keep
+            if current_result_diverged:
+                last_diverged_index = num_lines_to_keep
+                last_diverged_result = self.unique_stack_traces.copy()
+
+        raise Exception("Something went wrong with the search")
+
+    def reportResults(self, results, diverged_stack_frame_num):
+        """
+        Reports the final results
+        """
+        print('\nUnique Stack Traces - frames_kept:', diverged_stack_frame_num)
+        for trace, count in results.items():
+            print(f'{"*"*80}\nCount: {len(count)}\n')
+            if len(count) < 10:
+                print("\n".join(count))
```

Comment:
just print everything?

Reply:
What this logic does is print the hostnames of the outliers. If there are just a few (< 10) hosts in a "unique" stack trace, you'll see those hosts. Otherwise, you'll see just the number of hosts with that trace. It makes a lot of sense (e.g. There are 959 hosts in this stack trace, but host r1i2n3 PID 12345 is in this stace trace).